external/storm-kafka-new-consumer-api/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```
@@ -0,0 +1,445 @@
+        try {
+            kafkaConsumer.wakeup();
+            if(!consumerAutoCommitMode) {
+                commitOffsetsForAckedTuples();
+            }
+        } finally {
+            //remove resources
+            kafkaConsumer.close();
+        }
+    }
+
+    // TODO must declare multiple output streams
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        kafkaSpoutStreams.declareOutputFields(declarer);
+    }
+
+    // ====== Private helper methods ======
+
+    private void commitOffsetsForAckedTuples() {
+        final Map<TopicPartition, OffsetAndMetadata> nextCommitOffsets = new HashMap<>();
+
+        try {
+            for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+                final OffsetAndMetadata offsetAndMetadata = tpOffset.getValue().findNextCommitOffset();
+                if (offsetAndMetadata != null) {
+                    nextCommitOffsets.put(tpOffset.getKey(), offsetAndMetadata);
+                }
+            }
+
+            if (!nextCommitOffsets.isEmpty()) {
+                kafkaConsumer.commitSync(nextCommitOffsets);
+                LOG.debug("Offsets successfully committed to Kafka [{}]", nextCommitOffsets);
+                // Instead of iterating again, we could commit and update state for each TopicPartition in the prior loop,
+                // but the multiple network calls should be more expensive than iterating twice over a small loop
+                for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+                    final OffsetEntry offsetEntry = tpOffset.getValue();
+                    offsetEntry.updateAckedState(nextCommitOffsets.get(tpOffset.getKey()));
+                }
+            } else {
+                LOG.trace("No offsets to commit. {}", toString());
+            }
+        } catch (Exception e) {
+            LOG.error("Exception occurred while committing to Kafka offsets of acked tuples", e);
```

This is especially concerning because findNextCommitOffset says that it has side effects.  If we ignore an exception after calling something with side effects what will the results be?